[DF][RDatasetSpec] Minimal set of initial functionality

ikabadzhov opened the pull request:
* Entry range: if Start > End, logical error is thrown.
* Entry range: if Start == End, no work done
* Handle friends

ikabadzhov committed:
[DF][RDatasetSpec] Handle cases when Start >= End

If Start > End, logical error is thrown.
If Start == End, empty range is always taken.

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -41,14 +42,20 @@ struct RDatasetSpec {
-        fEndEntry(endEntry), fDefaultColumns(defaultColumns), fSubTreeNames(subTreenames)
+        fEndEntry((endEntry >= startEntry)
+                     ? endEntry
+                     : throw std::logic_error("RDatasetSpec: fEndEntry cannot be larger than fStartEntry.")),
```
> RDatasetSpec: fEndEntry cannot be larger than fStartEntry.

On the contrary, the end must be larger than the start

vepadulano replied:
Also, the condition you wrote is with `>=`, we need to decide if the `=` part is valid or not. In principle it can since the `TTreeReader` won't care, but logically I feel like it shouldn't

ikabadzhov replied:
I have to fix the bug with specifying `Range(X,X)` anyway. Then we can decide if we want this support in `RDatasetSpec` or not

ikabadzhov replied:
https://github.com/root-project/root/issues/10540

ikabadzhov committed:
[DF][RDatasetSpec] Remove the list of default columns

The RDatasetSpec constructor is no longer interacting with the
RLoopManager's fDefaultColumns. Now, all columns are always taken.

ikabadzhov committed:
[DF][RDatasetSpec] Avoid specifying chain name

ikabadzhov committed:
[DF][RDatasetSpec] Error out when fEnd > fBegin > NumEntries (in non-chains)

ikabadzhov committed:
[DF][RDatasetSpec] Handle case when # trees != # files

ikabadzhov committed:
[DF][RDatasetSpec] Initial version of friend trees handling

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,96 @@
+
+   void AddFriend(const std::string &treeName, const std::string &fileName)
+   {
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
```
```suggestion
// TODO: user might specify alias, empty for now
      fFriendInfo.fFriendNames.emplace_back(treeName, ""); 
```

ikabadzhov replied:
I agree. This is better. I kept it currently differently, so that the current `AddFriend` overloads are "matching" the overloads of the spec, e.g. specifying  either 1. 1tree/1file; 2. 1tree/Nfiles; 3. Ntrees/Nfiles.

So I think we should settle down better overloads for `AddFriend` and I would adapt these vector emplacements accordingly.

I recall one of the overloads proposed was:
```cpp
void AddFriend(const std::string &treeName, const std::string &fileName, const std::string &treeAlias = "")
```
Does it makes sense to simply add `const std::string &treeAlias = ""` to each `AddFriend` overload?

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,96 @@
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
+      fFriendInfo.fFriendFileNames.emplace_back(std::vector<std::string>{fileName});
+      fFriendInfo.fFriendChainSubNames.emplace_back(std::vector<std::string>{}); // this is a tree
```
```suggestion
      fFriendInfo.fFriendChainSubNames.emplace_back();
```

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,96 @@
-        fEndEntry(endEntry), fDefaultColumns(defaultColumns), fSubTreeNames(subTreenames)
+   void AddFriend(const std::string &treeName, const std::vector<std::string> &fileNames)
    {
+      fFriendInfo.fFriendNames.emplace_back(
```
Same as above

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,96 @@
+      fFriendInfo.fFriendNames.emplace_back(
+         treeName, treeName); // TODO: user might specify alias, now it is the tree name by default
+      fFriendInfo.fFriendFileNames.emplace_back(fileNames);
+      fFriendInfo.fFriendChainSubNames.emplace_back(std::vector<std::string>{treeName}); // now this is a chain
```
Since this is a TChain, the current element of the `fFriendChainSubNames` vector must be a vector with the same length of the corresponding element in the `fFriendFileNames` vector.

In practice, this should be a vector with length equal to `len(fileNames)` and each element of the vector is a repetition of `treeName`.

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,96 @@
-        fDefaultColumns(defaultColumns), fSubTreeNames(subTreenames)
+   void AddFriend(const std::vector<std::string> &treeNames, const std::vector<std::string> &fileNames)
    {
+      if (fileNames.size() != treeNames.size() && treeNames.size() != 1)
+         throw std::logic_error("RDatasetSpec's friend exepcts either N trees and N files, or 1 tree and N files.");
```
```suggestion
      if (fileNames.size() != treeNames.size())
         throw std::logic_error("The input treeNames and fileNames vectors must be of the same size.");
```

ikabadzhov replied:
Alright, a small note is that this gives "less" freedom to the user. And for us handling this does not come with extra cost. I have no idea of the potential use case, so I am going to apply the change.

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,96 @@
    {
+      if (fileNames.size() != treeNames.size() && treeNames.size() != 1)
+         throw std::logic_error("RDatasetSpec's friend exepcts either N trees and N files, or 1 tree and N files.");
+      fFriendInfo.fFriendNames.emplace_back(
+         treeNames[0], treeNames[0]); // TODO: user might specify alias, now it is the FIRST tree name by default
```
Here it is trickier, because the user is specifying explicitly a list of tree names it probably means all files in the chain have different filenames. Thus, it is not appropriate to give the TChain a name like this. I would argue this should just become
```suggestion
      fFriendInfo.fFriendNames.emplace_back();
```

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RLoopManager.hxx:
```diff
@@ -131,6 +131,8 @@ class RLoopManager : public RNodeBase {
    std::vector<ROOT::RDF::SampleCallback_t> fSampleCallbacks;
    RDFInternal::RNewSampleNotifier fNewSampleNotifier;
    std::vector<ROOT::RDF::RSampleInfo> fSampleInfos;
+   ROOT::Internal::TreeUtils::RFriendInfo fFriendInfo{};
+   std::vector<std::unique_ptr<TTree>> fFriends; ///< Friends of the tree/chain, if present
```
I don't understand why we need all these extra data members, especially the last one. The data members that belong to the spec should be removed, instead RLoopManager holds a RDataSetSpec as we said. I'm unsure about the vector of pointers. I am scared that we need it only because TChain doesn't keep alive the friends even after you attach them to it

ikabadzhov replied:
I realize now that `fFriendInfo` is not supposed to be there. It is true that I can be getting that from the `fSampleInfos` - I am applying this small change in a bit.

Regarding the vector of pointers - yes, it is exactly lifetime issues. If I don't have this vector => I cannot add the friends inside of the RLoopManager ctor. The alternative to keep them only inside `RunTreeReader` and `RunTreeProcessorMT`.

ikabadzhov replied:
Now I recall why I added `fFriendInfo` in the RLoopManager. I saw that we still added:
```cpp
   ULong64_t fStartEntry{0};
   ULong64_t fEndEntry{std::numeric_limits<ULong64_t>::max()};
``` 

I support the argument of redundancy, so I think we do not need `fStartEntry` and `fEndEntry` as they are already present in the RDatasetSpec, right?

vepadulano commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -377,13 +376,46 @@ RLoopManager::RLoopManager(const ROOT::RDF::RDatasetSpec &spec)
    }
    SetTree(chain);
+
+   const auto &friendNames = fFriendInfo.fFriendNames;
```
We need all the extra logic that attaches the friend info to the current chain that RDF is processing because previously we let the user first build the TTree+friends on their own and now they only specify it through the dataset spec. Right?

ikabadzhov replied:
this statement sounds right, yes. So it suffices to take care of the friend creation in the RLoopManager ctor.

vepadulano commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -501,12 +533,13 @@ void RLoopManager::RunTreeProcessorMT()
 /// Run event loop over one or multiple ROOT files, in sequence.
 void RLoopManager::RunTreeReader()
 {
+   if (fEndEntry == fStartEntry) // empty range => no work needed
```
When does this happen? The default values are 0 and max

vepadulano replied:
Is this only there if the user explicitly writes e.g. `{0, 0}` to the RDatasetSpec constructor?

ikabadzhov replied:
well, `fEndEntry` and `fStartEntry` are taken from the spec.

On the one hand, there might exist an empty TTree => no work.

On the other hand, current implementation of `TTreeReader` would error out if an empty range is provided. (https://github.com/root-project/root/issues/10540)

I currently do the first option, e.g. explicity no work.

The reason why I believe we need this extra check is: if I remove this case handling, in the subsequent lines `SetEntriesRange(fStartEntry, fEndEntry)` will take all entries.

So what do we want?
1. Do no work? --> current
2. Do all work? --> if I remove the `if (fEndEntry == fStartEntry)`
3. Error out? --> simply throw when  fEndEntry == fStartEntry

vepadulano commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -501,12 +533,13 @@ void RLoopManager::RunTreeProcessorMT()
-      r.SetEntriesRange(fStartEntry, fEndEntry);
-   }
+   if (r.SetEntriesRange(fStartEntry, fEndEntry)) // returning > 0 indicates invalid fStartEntry
+      throw std::runtime_error("RLoopManager: fStartEntry cannot be larger than the number of entries.");
```
Are we sure this is the only case of invalid start entry for TTreeReader?

ikabadzhov replied:
afaik yes. Because in case `fStartEntry > fEndEntry` I would have errored out in ctor. Moreover, if the range is `fStartEntry < "actual last entry" < fEndEntry` then `SetEntriesRange` returns 0 and will output the entries starting from the `fStartEntry` up to the the actual last entry.

ikabadzhov replied:
doublechecking https://github.com/root-project/root/blob/a402af862ace9573193500bcccaa8b4d56cc6632/tree/treeplayer/src/TTreeReader.cxx#L411 this looks like the only invalid thing that can go wrong at this stage.

ikabadzhov committed:
[DF][RDatasetSpec] Add aliases to the friends; Expect pairs in the third AddFriend constructor

ikabadzhov committed:
[DF][RDatasetSpec] Remove FriendInfo attribute from the RLoopManager

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -79,30 +80,34 @@ struct RDatasetSpec {
+      std::transform(begin(treeAndFileNames), end(treeAndFileNames), std::back_inserter(treeNames),
+                     [](auto const &pair) { return std::move(pair.first); });
+      std::vector<std::string> fileNames;
+      treeNames.reserve(treeAndFileNames.size());
```
```suggestion
      fileNames.reserve(treeAndFileNames.size());
```

ikabadzhov replied:
thanks, resolved now :)

ikabadzhov committed:
[DF][RDatasetSpec] Better pair to vectors conversion

Co-authored-by: Bernhard Manfred Gruber <bernhardmgruber@gmail.com>

ikabadzhov committed:
[DF][RDatasetSpec] Add Friends unit tests.

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RLoopManager.hxx:
```diff
@@ -118,7 +118,7 @@ class RLoopManager : public RNodeBase {
    std::shared_ptr<TTree> fTree{nullptr};
    const ColumnNames_t fDefaultColumns;
    ULong64_t fStartEntry{0};
-   ULong64_t fEndEntry{0};
+   ULong64_t fEndEntry{std::numeric_limits<ULong64_t>::max()};
```
The only new data member should be `fDatasetSpec`. Also these two need to go

ikabadzhov replied:
This was our initial idea, but I realized it also makes sense to keep these 2 attributes.

Reason why I kept `fStartEntry` and `fEndEntry` is that the RLoopManager constructor accepts an `RDatasetSpec` as an argument and handles all other specifications inside of the constructor already.

So the internal chain is already handled, and so are the friends. Why would I need to store strings of file and tree names, when I have already constructed everything?

Remark is that, even if I provide a spec as a member, I still need a vector of TTrees to hold the friends.

vepadulano replied:
I'm on the fence about this. on the one hand, it avoids all the issues that we were discussing about creating a spec, passing it to RDF, then changing the spec by adding friends. In the docs, we can say that the information about the dataset is only handled in the constructor then forgotten. On the other hand, maybe there are cases where we want to have quick access to the list of file names passed by the user. Currently, we would need to retrieve them from the internal TChain. By keeping the RDatasetSpec as data member we could just get the filenames from it.

ikabadzhov replied:
Sounds good. ~~I am adapting the RLoopManager now.~~

ikabadzhov replied:
hmm, I am thinking about it. I think I can come up with valid counter arguments:

I tend to agree with the first argument - especially that we should put that clearly in the docs. But I think whether the spec is a method or not, this snippet would be "equally bad". Right?
```
spec = RDatasetSpec("tree", "file")
df = RDataFrame(spec)
spec.AddFriend("fr1", "file1")
```

And regarding retrieving tree and file names: Hmm, the potential use case, sounds reasonable to me, but I see complications there. I am afraid that having the spec as a method might lead to more confusion. Because if the user want to retrieve this information, they would be able to, only if they constructed the RLoopManager with RDatasetSpec. Otherwise, we would need to adapt existing methods to also update the spec method of the RLoopManager.
I think it suffices to have public getters inside of the RDatasetSpec which fetch that info upon request. User only needs to recall which spec belonged to their dataframe.

vepadulano commented on tree/dataframe/inc/ROOT/RDF/RDatasetSpec.hxx:
```diff
@@ -13,44 +13,100 @@
     * contains its own tree with a different name from the global name of the dataset.
     */
-   std::vector<std::string> fSubTreeNames{};
+   std::vector<std::string> fTreeNames{};
+
+   /**
+    * A list of file names.
+    * They can contain the globbing characters supported by TChain. See TChain::Add for more information.
+    */
+   std::vector<std::string> fFileNameGlobs{};
```
Exactly like `RFriendInfo`, also `RDataSetSpec` should have three data members: one string with the name of the tree, one vector with the filenames and one vector with the (optional) names of trees in the files. The constructors should behave similarly to the overloads of `AddFriend` when filling these data members

ikabadzhov replied:
A question in the process of implementing. This optional argument seems applicable only for the third overload, of different tree names. Right?

ikabadzhov replied:
Alternative is to call the chain the same as the tree whenever I have only 1 tree name. And have no chain name, when I have different tree names in the cahin. this version is closer to our previous conclusion.

vepadulano commented on tree/dataframe/src/RLoopManager.cxx:
```diff
@@ -360,15 +360,14 @@ RLoopManager::RLoopManager(std::unique_ptr<RDataSource> ds, const ColumnNames_t
 {
    // A TChain has a global name
-   auto chain = std::make_shared<TChain>(spec.fDatasetName.c_str());
+   auto chain = std::make_shared<TChain>(spec.fTreeNames[0].c_str()); // use the first tree name (FOR NOW)
```
This should be constructed with `spec.fTreeName`

vepadulano replied:
Even better called `spec.fDatasetName` as before

ikabadzhov replied:
Commented above ^ **(https://github.com/vepadulano/root/pull/3#discussion_r892145107)**

ikabadzhov replied:
Current solution, which I considered (comming in a commit):
```diff
diff --git a/tree/dataframe/src/RLoopManager.cxx b/tree/dataframe/src/RLoopManager.cxx
index 4f560dcf3b..64c53f6933 100644
--- a/tree/dataframe/src/RLoopManager.cxx
+++ b/tree/dataframe/src/RLoopManager.cxx
@@ -364,24 +364,25 @@ RLoopManager::RLoopManager(const ROOT::RDF::RDatasetSpec &spec)
      fLoopType(ROOT::IsImplicitMTEnabled() ? ELoopType::kROOTFilesMT : ELoopType::kROOTFiles),
      fNewSampleNotifier(fNSlots), fSampleInfos(fNSlots)
 {
-   // A TChain has a global name
-   auto chain = std::make_shared<TChain>(spec.fTreeNames[0].c_str()); // use the first tree name (FOR NOW)
-
    if (spec.fTreeNames.size() == 1) {
+      // A TChain has a global name, that is the name of single tree
+      auto chain = std::make_shared<TChain>(spec.fTreeNames[0].c_str());
       // The global name of the chain is also the name of each tree in the list
       // of files that make the chain.
       for (const auto &f : spec.fFileNameGlobs)
          chain->Add(f.c_str());
+      SetTree(chain);
    } else {
       // Some other times, each different file has its own tree name, we need to
       // reconstruct the full path to the tree in each file and pass that to
       // TChain::Add
+      auto chain = std::make_shared<TChain>();
       for (auto i = 0u; i < spec.fFileNameGlobs.size(); i++) {
          const auto fullpath = spec.fFileNameGlobs[i] + "?#" + spec.fTreeNames[i];
          chain->Add(fullpath.c_str());
       }
+      SetTree(chain);
    }
-   SetTree(chain);
```
(I am yet to move the spec as a member)

vepadulano commented on tree/dataframe/test/dataframe_datasetspec.cxx:
```diff
@@ -194,36 +202,152 @@ TEST(RDFDatasetSpec, MultipleFiles)
    gSystem->Exec("rm file0.root file1.root file2.root");
 }
+
+TEST(RDatasetSpecTest, Friends)
```
Nice that you added these tests with friend trees, but I think it's a bit too convoluted for a single test. Maybe you can have a smaller example before this one and add a table representation of the dataset as a comment in this long one

ikabadzhov committed:
[DF][RDatasetSpec] Use std::move for vector of strings in RDatasetSpec

ikabadzhov committed:
[DF][RDatasetSpec] Update RDatasetSpec third constructor overload; Minor update in AddFriend handling

ikabadzhov committed:
[DF][RDatasetSpec] Better Chain naming

ikabadzhov committed:
[DF][RDatasetSpec] Make the friend test clearer

ikabadzhov committed:
[DF][RDatasetSpec] Spec copy passed; Use std::move where possible; Small improvement in AddFriend

ikabadzhov committed:
[DF][RDatasetSpec] Added various tests; Minor improvements

ikabadzhov commented:
Discussion is moved to the newer PR (directly to master)!
